src/components/documents/SelfCheck/SelfCheckTaskState.tsx

```diff
@@ -0,0 +1,52 @@
+import React from 'react';
+import { observer } from 'mobx-react-lite';
+import { useFirstMainDocument } from '@tdev-hooks/useFirstMainDocument';
+import Loader from '@tdev-components/Loader';
+import TaskState from '@tdev-components/documents/TaskState';
+import { MetaInit, TaskMeta } from '@tdev-models/documents/TaskState';
+import { ModelMeta as SolutionModelMeta } from '@tdev-models/documents/Solution';
+import { useDocumentRoot } from '@tdev-hooks/useDocumentRoot';
+import { NoneAccess } from '@tdev-models/helpers/accessPolicy';
+import { SelfCheckStateType } from '@tdev-components/documents/SelfCheck/models';
+import { SelfCheckContext } from '@tdev-components/documents/SelfCheck/shared';
+
+interface Props extends MetaInit {
+    includeQuestion: boolean;
+}
+
+const SelfCheckTaskState = observer(({ includeQuestion = true, pagePosition }: Props) => {
+    const context = React.useContext(SelfCheckContext);
+    if (!context) {
+        throw new Error('SelfCheckTaskState must be used within a SelfCheck');
+    }
+
+    const [taskMeta] = React.useState(new TaskMeta({ pagePosition }));
+    const [solutionMeta] = React.useState(new SolutionModelMeta({}));
+    const taskDoc = useFirstMainDocument(context.taskStateId, taskMeta, false);
+    const solutionDoc = useFirstMainDocument(context.solutionId, solutionMeta, false);
+    const solutionDocRoot = useDocumentRoot(context.solutionId, solutionMeta, false);
+
+    if (!taskDoc) {
+        return <Loader />;
+    }
+
+    const solutionAvailable = !!solutionDoc && !NoneAccess.has(solutionDocRoot.permission);
+
+    const states = [
+        SelfCheckStateType.Open,
+        includeQuestion ? SelfCheckStateType.Question : null,
+        solutionAvailable ? SelfCheckStateType.Reviewing : SelfCheckStateType.WaitingForSolution,
+        solutionAvailable ? SelfCheckStateType.Done : null
+    ].filter((state) => !!state);
+
+    if (solutionAvailable && taskDoc.taskState === SelfCheckStateType.WaitingForSolution) {
```

Comment:
This is a limitation of the current implementation: this side effect (and the second one just below) toggles the state between `WaitingForSolution` and `Reviewing`, depending on the availability of the solution. However, this only considers the availability for the current user. This means that, if a teacher now adds e.g. a user permission for a given user, this state change will only be triggered once that user loads the page. Until then, the teacher still sees the old state.

Reply:
A promising approach seems to be the use of a side effect. We could choose to only persist the `Reviewing` state, query the solution document root for each user's permission, and derive the `WaitingForSolution` state if a user doesn't have at least RO permission on the solution.

I am giving this a try and will follow up with another PR if it pans out.